distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso

```diff
@@ -854,15 +854,12 @@ type Table
             _ : Date_Range -> Error.throw (Unsupported_Database_Operation.Error "Cannot use `Date_Range` for `set` in the database.")
             _ -> Error.throw (Illegal_Argument.Error "Unsupported type for `Table.set`.")
 
-        name_is_generated column = case column of
-            _ : Column -> False
-            _ -> True
-
         ## If `new_name` was specified, use that. Otherwise, if `column` is a
            `Column`, use its name. In these two cases, do not make it unique.
-           Otherwise, make it unique.
+           Otherwise, make it unique. If set_mode is Update, however, do not
+           make it unique.
         new_column_name = if new_name != "" then new_name else
-            if name_is_generated column then unique.make_unique resolved.name else resolved.name
+            if column.is_a Column || set_mode==Set_Mode.Update then resolved.name else unique.make_unique resolved.name
```

Comment:
What about `Add_Or_Update`??

I think in this mode we should also _not_ uniquify the column name. But currently we do if I understand correctly.

```suggestion
            if column.is_a Column || set_mode==Set_Mode.Update || set_mode==Set_Mode.Add_Or_Update then resolved.name else unique.make_unique resolved.name
```

And then we should also add another test case, analogous to the one for Update, for Add_Or_Update: 
```
# set_mode=Set_Mode.Add_Or_Update is the default
t3 = t.set (Column_Operation.Add (Column_Ref.Name "X") 1)
t3.column_names . should_equal ["X"]
t3.at "X" . to_vector . should_equal [2, 3, 4]
```

Reply:
Done. But the test is actually this:

```
            t = table_builder [["X", [1, 2, 3]], ["[X] + 1", [10, 20, 30]]]
            # set_mode=Set_Mode.Add_Or_Update is the default
            t2 = t.set (Column_Operation.Add (Column_Ref.Name "X") 1)
            t2.column_names . should_equal ["X", "[X] + 1"]
            t2.at "X" . to_vector . should_equal [1, 2, 3]
            t2.at "[X] + 1" . to_vector . should_equal [2, 3, 4]
```

Since in the `Add_Or_Update` case, the new column is given a derived name `"[X] + 1"` rather than re-using the original name `"X"` as in the `Update` case.

I was a little surprised that `Add_Or_Update` would generate a derived name while `Update` always re-uses the original column name. Was that intentional? I tried changing the logic to use the original name in both cases, but that broke a lot of things (since nested subexpressions rely on the derived names to work at all).

I'm not going to merge this now because I want to make sure this is the behavior we want.